File: test/test_cartesian_conformance.jl
@testsnippet CartesianConformanceSetup begin
    using Test
    using EarthSciDiscretizations
    using JSON
end

@testitem "Cartesian cross-language conformance" setup = [CartesianConformanceSetup] tags = [:conformance, :grid] begin
    HARNESS_DIR = joinpath(@__DIR__, "..", "tests", "conformance", "grids", "cartesian")
    FIXTURES = JSON.parsefile(joinpath(HARNESS_DIR, "fixtures.json"))
    REL_TOL = Float64(FIXTURES["tolerance"]["relative"])

    # Query points are 0-indexed (binding-neutral); Julia accessors are
    # 1-indexed. Convert at the test boundary so the golden remains shared.
    qp_1idx(qp) = Tuple(Int(x) + 1 for x in qp)

    function close_rel(a::Real, b::Real, tol::Real)
        scale = max(1.0, abs(a), abs(b))
        return abs(a - b) <= tol * scale
    end

    function build_grid(opts)
        kw = Dict{Symbol, Any}()
        if haskey(opts, "edges")
            kw[:edges] = [Vector{Float64}(Float64.(e)) for e in opts["edges"]]
        else
            haskey(opts, "nx") && (kw[:nx] = Int(opts["nx"]))
            haskey(opts, "ny") && (kw[:ny] = Int(opts["ny"]))
            haskey(opts, "nz") && (kw[:nz] = Int(opts["nz"]))
            if haskey(opts, "extent")
                kw[:extent] = [(Float64(e[1]), Float64(e[2])) for e in opts["extent"]]
            end
        end
        kw[:ghosts] = Int(get(opts, "ghosts", 0))
        return EarthSciDiscretizations.grids.cartesian(; kw...)
    end

    for fixture in FIXTURES["fixtures"]
        name = fixture["name"]
        golden = JSON.parsefile(joinpath(HARNESS_DIR, "golden", "$name.json"))

        @testset "fixture=$name" begin
            grid = build_grid(fixture["opts"])
            N = ndims(grid)
            @test N == golden["ndim"]
            @test prod([length(grid.widths[d]) for d in 1:N]) == golden["n_cells"]
            @test length(fixture["query_points"]) == length(golden["cell_centers"])

            for (k, qp) in enumerate(fixture["query_points"])
                idx1 = qp_1idx(qp)

                c = cell_centers(grid, idx1...)
                gc = golden["cell_centers"][k]
                for d in 1:N
                    @test close_rel(Float64(c[d]), Float64(gc[d]), REL_TOL)
                end

                w = cell_widths(grid, idx1...)
                gw = golden["cell_widths"][k]
                for d in 1:N
                    @test close_rel(Float64(w[d]), Float64(gw[d]), REL_TOL)
                end

                v = cell_volume(grid, idx1...)
                @test close_rel(Float64(v), Float64(golden["cell_volume"][k]), REL_TOL)

                # Neighbors: Julia returns a Dict keyed by (axis_1based, side);
                # golden emits a sorted list of {axis, side, index} 0-indexed.
                nbr_dict = neighbors(grid, idx1...)
                nbr_list = Vector{Dict{String, Any}}()
                for d in 1:N
                    for s in (-1, +1)
                        haskey(nbr_dict, (d, s)) || continue
                        nb = nbr_dict[(d, s)]
                        push!(nbr_list, Dict{String, Any}(
                            "axis" => d - 1,
                            "side" => s,
                            "index" => [Int(x) - 1 for x in nb],
                        ))
                    end
                end
                gnbrs = golden["neighbors"][k]
                @test length(nbr_list) == length(gnbrs)
                for (got, exp) in zip(nbr_list, gnbrs)
                    @test got["axis"] == Int(exp["axis"])
                    @test got["side"] == Int(exp["side"])
                    @test got["index"] == [Int(x) for x in exp["index"]]
                end

                # Scalar metrics
                @test close_rel(
                    Float64(metric_eval(grid, :volume, idx1...)),
                    Float64(golden["metric_volume"][k]),
                    REL_TOL,
                )
                @test close_rel(
                    Float64(metric_eval(grid, :jacobian, idx1...)),
                    Float64(golden["metric_jacobian"][k]),
                    REL_TOL,
                )
                width_names = (:dx, :dy, :dz)
                face_names = (:face_area_x, :face_area_y, :face_area_z)
                for d in 1:N
                    @test close_rel(
                        Float64(metric_eval(grid, width_names[d], idx1...)),
                        Float64(golden["metric_$(width_names[d])"][k]),
                        REL_TOL,
                    )
                    @test close_rel(
                        Float64(metric_eval(grid, face_names[d], idx1...)),
                        Float64(golden["metric_$(face_names[d])"][k]),
                        REL_TOL,
                    )
                end

                # Metric tensor (identity for cartesian)
                g = metric_eval(grid, :g, idx1...)
                gg = golden["metric_g"][k]
                for i in 1:N, j in 1:N
                    @test close_rel(Float64(g[i][j]), Float64(gg[i][j]), REL_TOL)
                end
            end
        end
    end
end

File: python/tests/test_cartesian_conformance.py
"""Cross-language conformance for the cartesian grid family.

Verifies that the Python accessor runtime produces accessor outputs that
match the committed Julia-reference golden within the documented
``docs/GRIDS_API.md`` §4.2 tolerance. Corpus lives at
``tests/conformance/grids/cartesian/``.
"""

from __future__ import annotations

import json
from pathlib import Path

import pytest

from earthsci_toolkit import grids

HARNESS_DIR = (
    Path(__file__).resolve().parents[2]
    / "tests"
    / "conformance"
    / "grids"
    / "cartesian"
)
FIXTURES_PATH = HARNESS_DIR / "fixtures.json"
GOLDEN_DIR = HARNESS_DIR / "golden"
_SPEC = json.loads(FIXTURES_PATH.read_text())
REL_TOL = float(_SPEC["tolerance"]["relative"])
_FIXTURE_NAMES = [f["name"] for f in _SPEC["fixtures"]]


def _close_rel(a: float, b: float, tol: float = REL_TOL) -> bool:
    scale = max(1.0, abs(a), abs(b))
    return abs(a - b) <= tol * scale


def _fixture(name: str) -> dict:
    for f in _SPEC["fixtures"]:
        if f["name"] == name:
            return f
    raise KeyError(name)


def _build_grid(opts: dict):
    kw: dict = {"dtype": opts.get("dtype", "float64"), "ghosts": int(opts.get("ghosts", 0))}
    if "edges" in opts:
        kw["edges"] = [list(map(float, e)) for e in opts["edges"]]
    else:
        if "nx" in opts:
            kw["nx"] = int(opts["nx"])
        if "ny" in opts:
            kw["ny"] = int(opts["ny"])
        if "nz" in opts:
            kw["nz"] = int(opts["nz"])
        if "extent" in opts:
            kw["extent"] = [(float(lo), float(hi)) for lo, hi in opts["extent"]]
    return grids.cartesian(**kw)


@pytest.mark.parametrize("fixture_name", _FIXTURE_NAMES)
def test_conformance_against_golden(fixture_name: str) -> None:
    fixture = _fixture(fixture_name)
    grid = _build_grid(fixture["opts"])
    golden = json.loads((GOLDEN_DIR / f"{fixture_name}.json").read_text())

    ndim = grid.ndim
    assert ndim == int(golden["ndim"])
    assert grid.n_cells == int(golden["n_cells"])
    assert len(fixture["query_points"]) == len(golden["cell_centers"])

    width_names = ("dx", "dy", "dz")
    face_names = ("face_area_x", "face_area_y", "face_area_z")

    for k, qp in enumerate(fixture["query_points"]):
        idx = tuple(int(x) for x in qp)

        c = grid.cell_centers(*idx)
        gc = golden["cell_centers"][k]
        for d in range(ndim):
            assert _close_rel(float(c[d]), float(gc[d])), (
                f"{fixture_name}: cell_centers[{d}] at qp[{k}]={idx}"
            )

        w = grid.cell_widths(*idx)
        gw = golden["cell_widths"][k]
        for d in range(ndim):
            assert _close_rel(float(w[d]), float(gw[d])), (
                f"{fixture_name}: cell_widths[{d}] at qp[{k}]={idx}"
            )

        v = grid.cell_volume(*idx)
        assert _close_rel(float(v), float(golden["cell_volume"][k])), (
            f"{fixture_name}: cell_volume at qp[{k}]={idx}"
        )

        # Neighbors: Python returns a dict keyed by (axis, side) with 0-based
        # axis indices. Serialize to the same sorted list-of-entries shape
        # the golden emits (axis-major, low side before high side).
        nbr_dict = grid.neighbors(*idx)
        nbr_list = []
        for d in range(ndim):
            for s in (-1, +1):
                if (d, s) in nbr_dict:
                    nbr_list.append({
                        "axis": d,
                        "side": s,
                        "index": list(int(x) for x in nbr_dict[(d, s)]),
                    })
        gnbrs = golden["neighbors"][k]
        assert len(nbr_list) == len(gnbrs), (
            f"{fixture_name}: neighbor count at qp[{k}]={idx}"
        )
        for got, exp in zip(nbr_list, gnbrs):
            assert got["axis"] == int(exp["axis"])
            assert got["side"] == int(exp["side"])
            assert got["index"] == [int(x) for x in exp["index"]]

        # Scalar metrics
        assert _close_rel(
            float(grid.metric_eval("volume", *idx)),
            float(golden["metric_volume"][k]),
        )
        assert _close_rel(
            float(grid.metric_eval("jacobian", *idx)),
            float(golden["metric_jacobian"][k]),
        )
        for d in range(ndim):
            assert _close_rel(
                float(grid.metric_eval(width_names[d], *idx)),
                float(golden[f"metric_{width_names[d]}"][k]),
            ), f"{fixture_name}: metric {width_names[d]} at qp[{k}]={idx}"
            assert _close_rel(
                float(grid.metric_eval(face_names[d], *idx)),
                float(golden[f"metric_{face_names[d]}"][k]),
            ), f"{fixture_name}: metric {face_names[d]} at qp[{k}]={idx}"

        g = grid.metric_eval("g", *idx)
        gg = golden["metric_g"][k]
        for i in range(ndim):
            for j in range(ndim):
                assert _close_rel(float(g[i][j]), float(gg[i][j])), (
                    f"{fixture_name}: metric g[{i}][{j}] at qp[{k}]={idx}"
                )
